### PR TITLE
BUG: slow down with p=np.inf in 0.17 cKDTree.query

### PR DESCRIPTION
**Pull request by rainwoodman**
np.inf was not properly special cased, and pow(x, inf)
[ very slow on most libm, apparently] is used.

### CODE REVIEW

**Comment by sturlamolden** [comment]
@rgommers or @ev-br This should be pulled.

File: scipy/spatial/ckdtree/src/query.cxx [hunk 1]
```diff
@@ -381,6 +381,12 @@ query_single_point(const ckdtree *self,
                 npy_float64 tmp = x[split_dim] - split; 
                 if(NPY_LIKELY(p == 2)) {
                     inf2->side_distances()[split_dim] = tmp * tmp;
+                } else
+                if(NPY_LIKELY(p == 1)) {
+                    inf2->side_distances()[split_dim] = dabs(tmp);
+                } else
+                if(NPY_LIKELY(ckdtree_isinf(p))) {
```
Comment (perimosocordiae) [inline scipy/spatial/ckdtree/src/query.cxx]: These branches look like they do the same thing. Why not combine the p == 1 and the p == inf cases?

Comment (rainwoodman) [inline scipy/spatial/ckdtree/src/query.cxx]: p == inf and p == 1 are conceptually different cases, even though it happens that they are doing the same thing in this part of the algorithm.

It's more of a styling choice. The compiler should be able to make a decision to merge them or not during optimization.

Comment (sturlamolden) [inline scipy/spatial/ckdtree/src/query.cxx]: And I have a PR that makes the casings on ```p```  compile time constants, so it really doesn't matter. And besides we should not try to be smarter than the C++ compiler. In Python it matters, but this is C++.

**Comment by rgommers** [comment]
Thanks @rainwoodman.

It would be useful to add a benchmark for ``p=inf`` - there isn't one present yet right?

**Comment by sturlamolden** [comment]
I have a patch that replaces the conditionals on ```p``` and ```raw_boxsize_data``` with compile-time constants, so I would preferably like this PR merged soon as possible. Otherwise I would have to duplicate this PR in my patch. That is unless it is possible to pull this commit into my branch. But my Git skills are not that adwanced

**Comment by Eric89GXL** [comment]
+1 for merge once the benchmark issue is sorted out

**Comment by rainwoodman** [comment]
It's about 3 lines. I will add the benchmark later of the afternoon /
evening when I have a solid chunk of time.

**Comment by rainwoodman** [comment]
I increase the coverage of the benchmark suite to periodic and multiple p, but had to get rid of the py KDTree benchmarks.

**Comment by rgommers** [comment]
Removing the ``KDTree`` benchmarks seems fine to me, since we're going to make ``KDTree`` an alias to ``cKDTree`` anyway.

File: benchmarks/benchmarks/spatial.py [hunk 2]
```diff
@@ -40,49 +40,46 @@ def time_build(self, mnr, cls_name):
         else:
             self.cls(self.data)
 
-
+LeafSize = [8, 128]
+BoxSize = [None, 1.0]
```
Comment (pv) [inline benchmarks/benchmarks/spatial.py]: `LEAF_SIZES`, `BOX_SIZES` (pep8 says constants are all caps).
Also, needs newline between class declaration.

**Comment by ev-br** [comment]
Note to self: after merging this, only backport the actual change, https://github.com/scipy/scipy/pull/5957/commits/558de6fe4323a127679c2ba1b9fbb84fcf571ca0, not the benchmarks.

**Comment by ev-br** [comment]
OK, no more comments for two weeks, merging. Thanks @rainwoodman , all.

**Comment by rainwoodman** [comment]
Thanks for merging!

**Pull request merged**